File: scripts/test_solutions/678_valid_parenthesis_string.js
var checkValidString = function(s) {
    let lo = 0, hi = 0;
    for (const c of s) {
        if (c === '(') { lo++; hi++; }
        else if (c === ')') { lo--; hi--; }
        else { lo--; hi++; }
        if (hi < 0) return false;
        lo = Math.max(lo, 0);
    }
    return lo === 0;
};

File: scripts/test_solutions/678_valid_parenthesis_string.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int lo = 0;
        int hi = 0;
        for (char c : s) {
            if (c == '(') {
                lo++;
                hi++;
            } else if (c == ')') {
                lo--;
                hi--;
            } else {
                lo--;
                hi++;
            }
            if (hi < 0) {
                return false;
            }
            lo = max(lo, 0);
        }
        return lo == 0;
    }
};
